src/platforms/esp32/components/avm_builtins/i2c_resource.c

```diff
@@ -0,0 +1,743 @@
+    VALIDATE_VALUE(clock_speed_hz, term_is_integer);
+
+    i2c_port_t i2c_num = I2C_NUM_0;
+    term peripheral = interop_kv_get_value(opts, ATOM_STR("\xA", "peripheral"), global);
+    if (!term_is_invalid_term(peripheral)) {
+        if (!term_is_integer(peripheral)) {
+            ESP_LOGE(TAG, "Invalid parameter: peripheral is not an integer");
+            RAISE_ERROR(BADARG_ATOM);
+        }
+        i2c_num = term_to_int32(peripheral);
+        if (i2c_num < 0 || i2c_num > I2C_NUM_MAX - 1) {
+            ESP_LOGE(TAG, "Invalid parameter: i2c_num out of range");
+            RAISE_ERROR(BADARG_ATOM);
+        }
+    }
+
+    //
+    // Initialize config
+    //
+
+    i2c_config_t conf;
+    memset(&conf, 0, sizeof(i2c_config_t));
+    conf.mode = I2C_MODE_MASTER;
+    conf.scl_io_num = term_to_int32(scl);
+    conf.sda_io_num = term_to_int32(sda);
+    conf.sda_pullup_en = GPIO_PULLUP_ENABLE;
+    conf.scl_pullup_en = GPIO_PULLUP_ENABLE;
+    conf.master.clk_speed = term_to_int32(clock_speed_hz);
+
+    esp_err_t err;
+    err = i2c_param_config(i2c_num, &conf);
+    if (UNLIKELY(err != ESP_OK)) {
+        ESP_LOGE(TAG, "Failed to initialize I2C parameters.  err=%i", err);
+        if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+            ESP_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+            return OUT_OF_MEMORY_ATOM;
+        }
+        // TODO breaks i2c::open/1 API
+        return create_error_tuple(ctx, term_from_int(err));
+    }
+
+    //
+    // Install the I2C driver
+    //
+
+    err = i2c_driver_install(i2c_num, I2C_MODE_MASTER, 0, 0, 0);
+    if (UNLIKELY(err != ESP_OK)) {
+        ESP_LOGE(TAG, "Failed to install I2C driver.  err=%i", err);
+        if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+            ESP_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+            return OUT_OF_MEMORY_ATOM;
+        }
+        // TODO breaks i2c::open/1 API
+        return create_error_tuple(ctx, term_from_int(err));
+    }
+
+    ESP_LOGI(TAG, "I2C driver installed using I2C port %i", i2c_num);
+
+    //
+    // allocate and initialize the Nif resource
+    //
+
+    struct I2CResource *rsrc_obj = enif_alloc_resource(i2c_resource_type, sizeof(struct I2CResource));
+    if (IS_NULL_PTR(rsrc_obj)) {
+        i2c_driver_delete(i2c_num);
+        ESP_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+    rsrc_obj->transmitting_pid = term_invalid_term();
+    rsrc_obj->i2c_num = i2c_num;
+
+    if (UNLIKELY(memory_ensure_free(ctx, TERM_BOXED_RESOURCE_SIZE) != MEMORY_GC_OK)) {
+        i2c_driver_delete(i2c_num);
+        enif_release_resource(rsrc_obj);
+        ESP_LOGW(TAG, "Failed to allocate memory: %s:%i.", __FILE__, __LINE__);
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+    term obj = enif_make_resource(erl_nif_env_from_context(ctx), rsrc_obj);
```

Comment:
That's really and issue with the enif_ API, not with this PR.  I mean, if you want to fix that API, then I'd suggest doing so in a separate PR.  As it stands, this PR is using the prescribed API, like it or not.

Reply:
@pguyot Do you have any input on this issue?

Reply:
Erlang/OTP's enif_make_resource aborts when there is a malloc issue. enif_make_resource mimics this behavior instead of corrupting data but it really should be called after a memory_ensure_free.

However, the way we do it here seems safe, as we run memory_ensure_free with TERM_BOXED_RESOURCE_SIZE first.

The pattern in this PR is coherent with what can be found elsewhere, so if we want to change this we may want to do it in a separate PR.

Likewise, we may want to fix port_ensure_available.

Reply:
ok, I resolve this comment since we are not going to crash when running out of memory.